docs/extras/simp.md

```diff
@@ -0,0 +1,60 @@
+# Simp #
+
+The `simp` tactic works by applying "a conditional term rewriting system" (I don't really know what this means and would be interested to know more), to try and simplify the goal. You can actually watch what it's doing (which to the untrained eye looks like trying to apply lots of lemmas tagged with simp until it either makes progress or stops) by writing `set_option trace.simplify true` in your code. If you do this in the below
+
+```lean
+namespace xena
+
+definition cong (a b m : ℤ) : Prop := ∃ n : ℤ, b - a = m * n
+
+notation a ` ≡ ` b ` mod ` m  := cong a b m 
+set_option trace.simplify true
+theorem cong_refl (m : ℤ) : ∀ a : ℤ, a ≡ a mod m :=
+begin
+intro a,
+unfold cong,
+existsi (0:ℤ),
+simp
+end
+
+end xena
+```
+
+you will discover that Lean seems to apply simp when you do an unfold as well (at least I am assuming that this is what's going on -- unfold is also underlined in green).
+
+If you only want to see what worked rather than all the things that didn't, you could try `set_option trace.simplify.rewrite true`.
+
+### Simp lemmas
+
+In case you want to train simp to use certain extra lemmas (for example because they're coming up again and again in your work) you can add new lemmas for yourself. For example in mathlib in `algebra/field.lean` we find the line
+
+```lean
+@[simp] theorem ne_zero (u : units α) : (u : α) ≠ 0
+```
+
+This lemma is then added to `simp`'s armoury. Note several things however.
+
+1) You can't just make a random theorem into a simp lemma. It has to be of a certain kind, the most important kinds being those of the form `A=B` and `A↔B`. Note however that if you want to add `fact` to `simp`'s weaponry, you can prove
```

Comment:
This wasn't always the case, but `simp` will allow arbitrary facts as simp lemmas, by using `iff_true_intro` to turn `A` into `A <-> true`. (`units.ne_zero` is an example of this.) However, it only does this if the lemma isn't already in the form A = B or A <-> B, which means that sometimes you have to write it explicitly, if it gets the wrong idea with the default interpretation. So for example to discharge obligations of the form `x = x`, you can't just write `x = x` as a simp lemma, because it reads it as `x ~> x` and will loop. Instead, it has to be written as `x = x <-> true` so that it rewrites an entire equation to true.